File: packages/vault-item-types/src/native.ts
/**
 * Projection onto the base native secret (ADR 0087 §3).
 *
 * `sealed_store::Entry` is line one plus a `key: value` trailer, and it is
 * what `opensesame pass`, the `pm-bridges` binaries, `crates/kdbx-bridge` and
 * ConnectionRef materialisation all already read. Every item type declares how
 * it lands there, so a community type is readable by the host plane the day it
 * is authored, with no host-plane code.
 *
 * The projection is total in both directions: a type with no secret field
 * projects an empty line one, and a readback keeps trailer keys the definition
 * does not claim rather than dropping them.
 */

import { isString } from "@opensesame/os-domain";
import { FIELD_TYPES } from "./catalogue.js";
import {
  type FieldDefinition,
  type FieldValue,
  type FieldValues,
  type ItemTypeDefinition,
  definitionFields,
} from "./schema.js";

export type NativeEntry = {
  /** Line one of the entry. Never contains a newline. */
  readonly secret: string;
  /** The remainder, newline-terminated `key: value` lines. */
  readonly trailer: string;
};

export type NativeReadback = {
  readonly values: FieldValues;
  /** Trailer keys the definition does not claim, kept verbatim. */
  readonly extra: Readonly<Record<string, string>>;
};

const TRAILER_KEY = /^[a-z][a-z0-9_-]*$/;

type TrailerLine = {
  readonly key: string;
  readonly part: string | undefined;
  readonly raw: string;
};

/**
 * Split `key: value` or `key.part: value`, exactly as `split_line` does in
 * `crates/vault-item-types`.
 *
 * Written as a split rather than one regex because a regex that also
 * constrained the part would drop a line like `a.b.c: v` outright, while the
 * Rust side keeps it as an unclaimed key. The two planes read the same file;
 * they have to keep the same things.
 */
function splitTrailerLine(line: string): TrailerLine | null {
  const colon = line.indexOf(":");
  if (colon <= 0) return null;
  const head = line.slice(0, colon);
  const dot = head.indexOf(".");
  const key = dot === -1 ? head : head.slice(0, dot);
  const part = dot === -1 ? undefined : head.slice(dot + 1);
  if (!TRAILER_KEY.test(key)) return null;
  const rest = line.slice(colon + 1);
  return { key, part, raw: rest.startsWith(" ") ? rest.slice(1) : rest };
}

/**
 * Line one holds no newline and the trailer is line-oriented, so a value that
 * contains one is escaped rather than allowed to split the entry. `pass` users
 * read these files by eye, so the escape is the familiar one.
 */
export function encodeValue(value: string): string {
  return value
    .replaceAll("\\", "\\\\")
    .replaceAll("\r", "\\r")
    .replaceAll("\n", "\\n");
}

export function decodeValue(value: string): string {
  let out = "";
  for (let index = 0; index < value.length; index += 1) {
    const char = value[index];
    if (char !== "\\") {
      out += char;
      continue;
    }
    const next = value[index + 1];
    if (next === "n") out += "\n";
    else if (next === "r") out += "\r";
    else if (next === "\\") out += "\\";
    else {
      out += char;
      continue;
    }
    index += 1;
  }
  return out;
}

function asText(value: FieldValue | undefined): string {
  return isString(value) ? value : "";
}

function asList(value: FieldValue | undefined): readonly string[] {
  if (Array.isArray(value)) return value.filter(isString);
  return isString(value) ? [value] : [];
}

function asRecord(value: FieldValue | undefined): Record<string, string> {
  if (value === undefined || isString(value) || Array.isArray(value)) return {};
  return value;
}

/**
 * `pass-otp` reads an `otpauth://` URI only from a line that *starts* with it
 * (`find_otpauth_in_trailer`), so a TOTP seed is written bare rather than
 * behind its trailer key. The exception belongs to the field type, not to any
 * definition — which is the point of the catalogue being closed.
 */
function isBareOtpauth(field: FieldDefinition, text: string): boolean {
  return field.type === "totp" && text.toLowerCase().startsWith("otpauth://");
}

function trailerLines(
  field: FieldDefinition,
  key: string,
  value: FieldValue | undefined,
): readonly string[] {
  const spec = FIELD_TYPES[field.type];
  if (spec.valueKind === "record") {
    const record = asRecord(value);
    return spec.parts
      .filter((part) => (record[part.id] ?? "") !== "")
      .map(
        (part) => `${key}.${part.id}: ${encodeValue(record[part.id] ?? "")}`,
      );
  }
  if (field.multiple === true) {
    return asList(value)
      .filter((entry) => entry !== "")
      .map((entry) => `${key}: ${encodeValue(entry)}`);
  }
  const text = asText(value);
  if (text === "") return [];
  if (isBareOtpauth(field, text)) return [text];
  return [`${key}: ${encodeValue(text)}`];
}

export function toNativeEntry(
  definition: ItemTypeDefinition,
  values: FieldValues,
): NativeEntry {
  const byId = new Map(definitionFields(definition).map((f) => [f.id, f]));
  const secretId = definition.spec.native.secret;
  const secret = secretId === null ? "" : encodeValue(asText(values[secretId]));

  const lines: string[] = [];
  for (const mapping of definition.spec.native.trailer) {
    const field = byId.get(mapping.field);
    if (field === undefined) continue;
    lines.push(...trailerLines(field, mapping.key, values[mapping.field]));
  }
  const trailer = lines.length === 0 ? "" : `${lines.join("\n")}\n`;
  return { secret, trailer };
}

/** Render the projection as the plaintext `sealed_store::Entry` body. */
export function renderNativeEntry(entry: NativeEntry): string {
  return `${entry.secret}\n${entry.trailer}`;
}

export function fromNativeEntry(
  definition: ItemTypeDefinition,
  entry: NativeEntry,
): NativeReadback {
  const byKey = new Map(
    definition.spec.native.trailer.map((mapping) => [
      mapping.key,
      mapping.field,
    ]),
  );
  const byId = new Map(definitionFields(definition).map((f) => [f.id, f]));

  const values: Record<string, FieldValue> = {};
  const extra: Record<string, string> = {};
  const lists = new Map<string, string[]>();
  const records = new Map<string, Record<string, string>>();

  const secretId = definition.spec.native.secret;
  if (secretId !== null) values[secretId] = decodeValue(entry.secret);

  const totpField = definition.spec.native.trailer.find(
    (mapping) => byId.get(mapping.field)?.type === "totp",
  );

  for (const line of entry.trailer.split("\n")) {
    if (line.trim() === "") continue;
    if (line.trim().toLowerCase().startsWith("otpauth://")) {
      if (totpField !== undefined) values[totpField.field] = line.trim();
      else extra.otpauth = line.trim();
      continue;
    }
    const parsed = splitTrailerLine(line);
    if (parsed === null) continue;
    const { key, part, raw } = parsed;
    const fieldId = byKey.get(key);
    const field = fieldId === undefined ? undefined : byId.get(fieldId);
    if (fieldId === undefined || field === undefined) {
      extra[part === undefined ? key : `${key}.${part}`] = decodeValue(raw);
      continue;
    }
    const spec = FIELD_TYPES[field.type];
    if (spec.valueKind === "record") {
      const known = spec.parts.some((candidate) => candidate.id === part);
      if (part === undefined || !known) {
        extra[part === undefined ? key : `${key}.${part}`] = decodeValue(raw);
        continue;
      }
      const record = records.get(fieldId) ?? {};
      record[part] = decodeValue(raw);
      records.set(fieldId, record);
      continue;
    }
    if (field.multiple === true) {
      const list = lists.get(fieldId) ?? [];
      list.push(decodeValue(raw));
      lists.set(fieldId, list);
      continue;
    }
    values[fieldId] = decodeValue(raw);
  }

  for (const [fieldId, list] of lists) values[fieldId] = list;
  for (const [fieldId, record] of records) values[fieldId] = record;
  return { values, extra };
}

File: crates/vault-item-types/src/native.rs
//! Projection onto the base native secret (ADR 0087 §3).
//!
//! `sealed_store::Entry` is line one plus a `key: value` trailer, and it is
//! what `opensesame pass`, the `pm-bridges` binaries and `kdbx-bridge` already
//! read. Because every item type declares how it lands there, a community type
//! is readable by the host plane the day it is authored, with no host-plane
//! code at all.
//!
//! `packages/vault-item-types/src/native.ts` is the same function; the two are
//! held together by `tests/conformance.rs`.

use std::collections::BTreeMap;

use opensesame_sealed_store::Entry;
use serde::{Deserialize, Serialize};

use crate::catalogue::{FieldShape, FieldTypeId};
use crate::schema::{FieldDefinition, ItemTypeDefinition};

/// A field value as stored: text, a list of text for a repeating field, or the
/// named parts of a record-shaped one. Text throughout, because the base
/// native secret is text and every type projects onto it.
#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
#[serde(untagged)]
pub enum FieldValue {
    Text(String),
    List(Vec<String>),
    Parts(BTreeMap<String, String>),
}

impl FieldValue {
    #[must_use]
    pub fn as_text(&self) -> &str {
        match self {
            Self::Text(text) => text,
            _ => "",
        }
    }

    #[must_use]
    pub fn as_list(&self) -> Vec<String> {
        match self {
            Self::Text(text) if text.is_empty() => Vec::new(),
            Self::Text(text) => vec![text.clone()],
            Self::List(list) => list.clone(),
            Self::Parts(_) => Vec::new(),
        }
    }

    #[must_use]
    pub fn as_parts(&self) -> BTreeMap<String, String> {
        match self {
            Self::Parts(parts) => parts.clone(),
            _ => BTreeMap::new(),
        }
    }
}

pub type FieldValues = BTreeMap<String, FieldValue>;

/// What a readback recovered: the definition's own fields, plus trailer keys
/// it does not claim, kept verbatim rather than dropped.
#[derive(Clone, Debug, Default, PartialEq, Eq)]
pub struct Readback {
    pub values: FieldValues,
    pub extra: BTreeMap<String, String>,
}

/// Line one holds no newline and the trailer is line-oriented, so a value that
/// contains one is escaped rather than allowed to split the entry.
#[must_use]
pub fn encode_value(value: &str) -> String {
    value
        .replace('\\', "\\\\")
        .replace('\r', "\\r")
        .replace('\n', "\\n")
}

#[must_use]
pub fn decode_value(value: &str) -> String {
    let mut out = String::with_capacity(value.len());
    let mut chars = value.chars().peekable();
    while let Some(current) = chars.next() {
        if current != '\\' {
            out.push(current);
            continue;
        }
        match chars.peek() {
            Some('n') => {
                out.push('\n');
                chars.next();
            }
            Some('r') => {
                out.push('\r');
                chars.next();
            }
            Some('\\') => {
                out.push('\\');
                chars.next();
            }
            _ => out.push(current),
        }
    }
    out
}

/// `pass-otp` reads an `otpauth://` URI only from a line that *starts* with
/// it, so a TOTP seed is written bare rather than behind its trailer key. The
/// exception belongs to the field type, not to any definition.
fn is_bare_otpauth(field: &FieldDefinition, text: &str) -> bool {
    field.field_type == FieldTypeId::Totp && text.to_ascii_lowercase().starts_with("otpauth://")
}

fn trailer_lines(field: &FieldDefinition, key: &str, value: Option<&FieldValue>) -> Vec<String> {
    let Some(value) = value else {
        return Vec::new();
    };
    if field.field_type.shape() == FieldShape::Record {
        let parts = value.as_parts();
        return field
            .field_type
            .parts()
            .iter()
            .filter_map(|part| {
                let text = parts.get(part.id)?;
                if text.is_empty() {
                    return None;
                }
                Some(format!("{key}.{}: {}", part.id, encode_value(text)))
            })
            .collect();
    }
    if field.repeats() {
        return value
            .as_list()
            .into_iter()
            .filter(|entry| !entry.is_empty())
            .map(|entry| format!("{key}: {}", encode_value(&entry)))
            .collect();
    }
    let text = value.as_text();
    if text.is_empty() {
        Vec::new()
    } else if is_bare_otpauth(field, text) {
        vec![text.to_owned()]
    } else {
        vec![format!("{key}: {}", encode_value(text))]
    }
}

/// Project an item's values onto the base native secret.
#[must_use]
pub fn to_entry(definition: &ItemTypeDefinition, values: &FieldValues) -> Entry {
    let secret = definition
        .spec
        .native
        .secret
        .as_deref()
        .and_then(|id| values.get(id))
        .map_or_else(String::new, |value| encode_value(value.as_text()));

    let mut lines: Vec<String> = Vec::new();
    for mapping in &definition.spec.native.trailer {
        let Some(field) = definition.field(&mapping.field) else {
            continue;
        };
        lines.extend(trailer_lines(
            field,
            &mapping.key,
            values.get(&mapping.field),
        ));
    }
    let trailer = if lines.is_empty() {
        String::new()
    } else {
        format!("{}\n", lines.join("\n"))
    };
    // Round-tripping through `Entry::parse` rather than building the struct
    // is what picks up the bare `otpauth://` line as the entry's OTP; building
    // it directly and calling `with_otp(None)` would strip that line back out.
    Entry::parse(&format!("{secret}\n{trailer}"))
}

/// Split a trailer line into its key, optional record part, and raw value.
fn split_line(line: &str) -> Option<(&str, Option<&str>, &str)> {
    let (head, raw) = line.split_once(':')?;
    if head.is_empty() || !head.is_ascii() {
        return None;
    }
    let (key, part) = match head.split_once('.') {
        Some((key, part)) => (key, Some(part)),
        None => (head, None),
    };
    let mut first = key.chars();
    if !first.next().is_some_and(|c| c.is_ascii_lowercase()) {
        return None;
    }
    if !key
        .chars()
        .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == '_' || c == '-')
    {
        return None;
    }
    Some((key, part, raw.strip_prefix(' ').unwrap_or(raw)))
}

/// Recover an item's values from a native entry. Total: trailer keys the
/// definition does not claim survive in `extra`.
#[must_use]
pub fn from_entry(definition: &ItemTypeDefinition, entry: &Entry) -> Readback {
    let by_key: BTreeMap<&str, &str> = definition
        .spec
        .native
        .trailer
        .iter()
        .map(|mapping| (mapping.key.as_str(), mapping.field.as_str()))
        .collect();
    let totp_field = definition
        .spec
        .native
        .trailer
        .iter()
        .find(|mapping| {
            definition
                .field(&mapping.field)
                .is_some_and(|field| field.field_type == FieldTypeId::Totp)
        })
        .map(|mapping| mapping.field.clone());

    let mut readback = Readback::default();
    if let Some(secret) = definition.spec.native.secret.as_deref() {
        readback.values.insert(
            secret.to_owned(),
            FieldValue::Text(decode_value(&entry.secret)),
        );
    }
    let mut lists: BTreeMap<String, Vec<String>> = BTreeMap::new();
    let mut records: BTreeMap<String, BTreeMap<String, String>> = BTreeMap::new();

    for line in entry.trailer.lines() {
        let trimmed = line.trim();
        if trimmed.is_empty() {
            continue;
        }
        if trimmed.to_ascii_lowercase().starts_with("otpauth://") {
            match totp_field.as_ref() {
                Some(field) => {
                    readback
                        .values
                        .insert(field.clone(), FieldValue::Text(trimmed.to_owned()));
                }
                None => {
                    readback
                        .extra
                        .insert("otpauth".to_owned(), trimmed.to_owned());
                }
            }
            continue;
        }
        let Some((key, part, raw)) = split_line(line) else {
            continue;
        };
        absorb_line(
            definition,
            &by_key,
            &mut readback,
            &mut lists,
            &mut records,
            (key, part, raw),
        );
    }

    for (field, list) in lists {
        readback.values.insert(field, FieldValue::List(list));
    }
    for (field, parts) in records {
        readback.values.insert(field, FieldValue::Parts(parts));
    }
    readback
}

fn absorb_line(
    definition: &ItemTypeDefinition,
    by_key: &BTreeMap<&str, &str>,
    readback: &mut Readback,
    lists: &mut BTreeMap<String, Vec<String>>,
    records: &mut BTreeMap<String, BTreeMap<String, String>>,
    line: (&str, Option<&str>, &str),
) {
    let (key, part, raw) = line;
    let extra_key = part.map_or_else(|| key.to_owned(), |part| format!("{key}.{part}"));
    let Some(field) = by_key.get(key).and_then(|id| definition.field(id)) else {
        readback.extra.insert(extra_key, decode_value(raw));
        return;
    };
    if field.field_type.shape() == FieldShape::Record {
        let known = part.is_some_and(|name| {
            field
                .field_type
                .parts()
                .iter()
                .any(|candidate| candidate.id == name)
        });
        match part.filter(|_| known) {
            Some(name) => {
                records
                    .entry(field.id.clone())
                    .or_default()
                    .insert(name.to_owned(), decode_value(raw));
            }
            None => {
                readback.extra.insert(extra_key, decode_value(raw));
            }
        }
        return;
    }
    if field.repeats() {
        lists
            .entry(field.id.clone())
            .or_default()
            .push(decode_value(raw));
        return;
    }
    readback
        .values
        .insert(field.id.clone(), FieldValue::Text(decode_value(raw)));
}
